### PR TITLE
My first class

### PR DESCRIPTION
**Pull request by ClearFly22**

### CODE REVIEW

**Comment by muramura** [comment]
@ClearFly22 san....
I am aware that this PR is improperly up.
If you really want to PR, please describe the title and description.

File: ArduCopter/mode.h [hunk 1]
```diff
@@ -1191,3 +1191,77 @@ class ModeZigZag : public Mode {
...
 };
+
+class ModeHoldUltra : public ModeGuided {
+
+public:
+
+    // inherit constructor
+    using Copter::ModeGuided::Mode;
+
+	// Init und run werden überschrieben:
```
Comment (amilcarlucas) [inline ArduCopter/mode.h]: No comments in german, please use english

File: ArduCopter/mode.h [hunk 2]
```diff
@@ -1191,3 +1191,77 @@ class ModeZigZag : public Mode {
...
+    bool init(bool ignore_checks) override;
+    void run() override;
+
+	// Punkte, die erforderlich sind
```
Comment (amilcarlucas) [inline ArduCopter/mode.h]: No german please

File: ArduCopter/mode_hold_ultra.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
+#include "Copter.h"
+
+/*
+ * Init and run calls for stabilize flight mode
+ */
+
+// stabilize_init - initialise stabilize controller
+bool Copter::ModeHoldUltra::init(bool ignore_checks)
+{
+    // Ronny: Hier kann ich die Klasse initialisieren
```
Comment (amilcarlucas) [inline ArduCopter/mode_hold_ultra.cpp]: No names, and no german

**Comment by ClearFly22** [comment]
Hi amilcarlucas 

Thanks, done.

I see in your code, there are only two ultrasonic sensors supported. Why?
You habe never try to build a hold mode for indoor with ultrasonic navigation?
Or know you a better solution?

Greetings from Austria
Ronald Lilleg



Gesendet von Mail für Windows 10

Von: Amilcar Lucas
Gesendet: Samstag, 2. Februar 2019 02:34
An: ArduPilot/ardupilot
Cc: Ronald Lilleg; Mention
Betreff: Re: [ArduPilot/ardupilot] My first class (#10392)

@amilcarlucas requested changes on this pull request.

In ArduCopter/mode.h:
> @@ -1191,3 +1191,77 @@ class ModeZigZag : public Mode {

 

     uint32_t reach_wp_time_ms = 0;  // time since vehicle reached destination (or zero if not yet reached)

 };

+

+class ModeHoldUltra : public ModeGuided {

+

+public:

+

+    // inherit constructor

+    using Copter::ModeGuided::Mode;

+

+	// Init und run werden überschrieben:

No comments in german, please use english

In ArduCopter/mode.h:
> @@ -1191,3 +1191,77 @@ class ModeZigZag : public Mode {

 

     uint32_t reach_wp_time_ms = 0;  // time since vehicle reached destination (or zero if not yet reached)

 };

+

+class ModeHoldUltra : public ModeGuided {

+

+public:

+

+    // inherit constructor

+    using Copter::ModeGuided::Mode;

+

+	// Init und run werden überschrieben:

+    bool init(bool ignore_checks) override;

+    void run() override;

+

+	// Punkte, die erforderlich sind

No german please

In ArduCopter/mode_hold_ultra.cpp:
> @@ -0,0 +1,76 @@

+#include "Copter.h"

+

+/*

+ * Init and run calls for stabilize flight mode

+ */

+

+// stabilize_init - initialise stabilize controller

+bool Copter::ModeHoldUltra::init(bool ignore_checks)

+{

+    // Ronny: Hier kann ich die Klasse initialisieren

No names, and no german

**Comment by amilcarlucas** [comment]
According to https://github.com/ArduPilot/ardupilot/commit/5eff01a86f9381c6f5a5f9a6a6c38623c409ccd4 the ardupilot code now support up-to 10 range finders.

File: ArduCopter/mode_hold_ultra.cpp (hunk 3 above)
Comment (ClearFly22) [reply ArduCopter/mode_hold_ultra.cpp]: Thanks for 10 range finders !!!

**Comment by rmackay9** [comment]
This is interesting work and the developer has put some good effort in.  I think the flight modes are not general purpose enough for us to merge to master though.  I think once we have Lua scripts able to set attitude and position targets these are the types of custom flight modes that we hope developers can create using Lua.

 The HC range finder driver is also interesting because people keep buying them (probably partially because they are so cheap).  I think though that we have decided in the past that we won't support these sonar because they put too much pressure on the flight controller and don't actually work very well.

So well done on a good bit of development and I hope we see more contributions from this developer in the future

**Pull request closed without merging**